### ISSUE TITLE
Stop using Retrofit BaseUrl for switching REST api endpoint

### ISSUE THREAD

**Issue by artem-zinnatullin** [artem-zinnatullin/qualitymatters#111]
See https://github.com/square/retrofit/pull/1652

**Comment by vanniktech** [comment]
Retrofit2 is released :tada:

**Comment by lenguyenthanh** [comment]
For this situation, I'm using a [Configuration](https://github.com/lenguyenthanh/Snowball/blob/master/app%2Fsrc%2Fmain%2Fjava%2Fcom%2Flenguyenthanh%2Fsnowball%2Fapp%2Fconfig%2FConfiguration.java) class. Which can inject to ApiModule to get Retrofit a base url. 

With this, we can easy to change configuration environment for each build type as well as function testing. 

Interceptor is an interesting way, but I think using Configuration is more natural.

**Comment by artem-zinnatullin** [comment]
The thing is that I'd like to be able to change endpoint url without restarting the app, I think, recreation of Retrofit is the best option in such case.

See https://github.com/square/retrofit/pull/1652

**Comment by lenguyenthanh** [comment]
Change endpoint url without restart app is really cool, but recreation of Retrofit with clean code in mind is quite hard. Looking forward the solution for this.

**Comment by artem-zinnatullin** [comment]
Recreation can be done via Dagger, we'll either recreate only Retrofit or also objects that depend on it, not sure yet

**Comment by ArazAbishov** [comment]
@artem-zinnatullin This issue turns out to be a little bit more complicated that it seems to be. Even if you re-create the object graph (with new retrofit instance pointing to another URL), you need to propagate / re-inject dependencies into targets (views, presenters). Otherwise, behaviour of the app will be unpredictable. The most obvious, but not elegant solution is to restart process by using: https://github.com/JakeWharton/ProcessPhoenix. If process will be restarted, we won't have stale instances of dependencies within application. 

What do you think?

**Comment by artem-zinnatullin** [comment]
Noo, process restart is not an option because it won't work in UI tests
since they run in same process (instrumentation API).

**Comment by Yurii-Laguta** [comment]
@artem-zinnatullin would you accept Interceptor solution as per conversation in https://github.com/square/retrofit/pull/1652 ?
This issues blocks #208

**Comment by artem-zinnatullin** [comment]
@Yurii-Laguta yep, we can start with that, would be better to re-create Retrofit on base URL change though
